### ISSUE TITLE
[HOLD for payment 2023-03-21] [$1000] Change in the emoji skin tone in one platform (mweb) triggers or selects the two different skin tones on the other platform (android)

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#14644]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open android app with account user A and go to any chat and select emoji picker
2. Check for the skin tone which is currently selected
3. Now go to mweb and login with the same user account A and go to any chat and select emoji picker
4. Next try to select a different skin tone in the mweb
5. Go back to android app and notice that two skin tones are selected at the same time. The latest skin tone should have been selected.

## Expected Result:
Only the latest skin tone should have been selected

## Actual Result:
Both skin tones (previous and latest) are selected. In order to view the latest skin tone , we need to close the emoji picker and re-open it.

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome

**Version Number:** 1.2.61-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/215234442-b1311cba-b901-483f-8bc7-e15306dfc7bb.mp4

https://user-images.githubusercontent.com/43996225/215234447-18a7db72-17a2-4832-af40-70eab9e76c72.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @priya-zha
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1674839129166789

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01559df482f2b25fe9</li>
        <li>Upwork Job ID: 1624163663019192320</li>
        <li>Last Price Increase: 2023-03-07</li>
    </ul>
</details>

**Comment by mallenexpensify** [comment]
Android keeps crashing on me in BrowserStack, will try again tomorrow

**Comment by mallenexpensify** [comment]
I can't get skin tone to change accurately or consistently for me, dropped in convo here https://expensify.slack.com/archives/C049HHMV9SM/p1675437076355079

**Comment by MelvinBot** [comment]
@mallenexpensify Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by mallenexpensify** [comment]
I'm going to put this on hold pending 
- https://github.com/Expensify/App/issues/14876

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01559df482f2b25fe9

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor Plus for review of internal employee PR - @thesahindia (`Internal`)

**Comment by mallenexpensify** [comment]
Thanks for the thorough explanation @fedirjh 
@thesahindia can you review and provide feedback? I added the `Internal` label to assign you, reckon this can be made `External` if you aggree this and https://github.com/Expensify/App/issues/14876 are unrelated

**Comment by mallenexpensify** [comment]
@thesahindia it looks like https://github.com/Expensify/App/issues/14876 hit production.  Can you retest this to see if it's still happening, if so I'll take off hold (I'm back from being partially OOO tomorrow)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Luke9389 (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by mallenexpensify** [comment]
Pretty sure this can be made `External`, @Luke9389 , can you confirm?

**Comment by MelvinBot** [comment]
@mallenexpensify, @Luke9389, @thesahindia Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MelvinBot** [comment]
Triggered auto assignment to @francoisl (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by mallenexpensify** [comment]
@francoisl I'm pretty sure this can be external, can you confirm?

**Comment by francoisl** [comment]
Yes, going to apply the label and remove the `[HOLD]` from the title.

**Comment by MelvinBot** [comment]
Current assignee @mallenexpensify is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @thesahindia is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @francoisl is eligible for the External assigner, not assigning anyone new.

**Comment by BeeMargarida** [comment]
Hi, I'm Ana from Callstack. Is this being worked on? If not, I can pick it up

**Comment by MelvinBot** [comment]
📣 @BeeMargarida! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by MelvinBot** [comment]
📣 @BeeMargarida You have been assigned to this job by @mallenexpensify!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by mallenexpensify** [comment]
@BeeMargarida , as there are no existing proposals for the fix, I've assigned to you.

**Comment by BeeMargarida** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When changing the skin tone in one device, two skin tones in another device become selected (the previous one and the new one).

### What is the root cause of that problem?

There are two ways for an item to become highlighted: if it's being hovered on or it's selected. The problem is that when the skin tone is selected in another device, the highlighted index is not updated, so it becomes with the previously state it had. 

### What changes do you think we should make in order to solve the problem?

Add a `componentDidUpdate` that updates the `highlightedIndex` if the `props.preferredSkinTone` changed, with similar logic to what's already in the `componentDidMount`.

```
componentDidUpdate(prevProps) {
    if (prevProps.preferredSkinTone === this.props.preferredSkinTone) { return; }

    const selectedEmoji = getSkinToneEmojiFromIndex(this.props.preferredSkinTone);
    this.setState({highlightedIndex: selectedEmoji.skinTone});
}
```

### What alternative solutions did you explore? (Optional)

None 


The proposal is basically the same as the one above, after exploring solutions it seems like it is the optimal one.

**Comment by BeeMargarida** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

When changing the skin tone in one device, two skin tones in another device become selected (the previous one and the new one).

### What is the root cause of that problem?

There are two ways for an item to become highlighted: if it's being hovered on or it's selected. The problem is that when the skin tone is selected in another device, the highlighted index is not updated, so it becomes with the previously state it had. 

### What changes do you think we should make in order to solve the problem?

Add a `componentDidUpdate` that updates the `highlightedIndex` if the `props.preferredSkinTone` changed, with similar logic to what's already in the `componentDidMount`.

```
componentDidUpdate(prevProps) {
    if (prevProps.preferredSkinTone === this.props.preferredSkinTone) { return; }

    const selectedEmoji = getSkinToneEmojiFromIndex(this.props.preferredSkinTone);
    this.setState({highlightedIndex: selectedEmoji.skinTone});
}
```

### What alternative solutions did you explore? (Optional)

None

The proposal is very similar to the one above, after exploring solutions it seems like it is the optimal one.

https://user-images.githubusercontent.com/25725586/223693675-00166d8e-0802-4db2-b299-56f95e7c33f5.mov

**Comment by mallenexpensify** [comment]
Sorry @fedirjh 
1. It didn't start with `Proposal` so I didn't treat it as one. 
2. It was posted before the `External` or `Help Wanted` labels were added. I _thought_ we had language about this in CONTRIBUTING.md but I'm unable to find. I think it might have been removed when we started locking issues a while back. 

That said... I unlocked this issue and said. 
> Unlocked in case anyone has ideas on what's going on with skin tone issues

My intent wasn't to have proposals submitted but I didn't make that clear. 

@BeeMargarida please keep working on the issue, once code has been deployed to production let's compare what was used and how similar it is to what @fedirjh initially proposed.  At that point we'll discuss potential compensation.

**Comment by BeeMargarida** [comment]
@mallenexpensify Just to check, is my proposal accepted, can I move on to the code?

**Comment by mallenexpensify** [comment]
Checking internally @BeeMargarida, should have an update by tomorrow

https://expensify.slack.com/archives/C01SKUP7QR0/p1678388439242879

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.83-3 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15775


If no regressions arise, payment will be issued on 2023-03-21. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@thesahindia / @francoisl] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@thesahindia / @francoisl] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@thesahindia / @francoisl] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@mallenexpensify] Determine if we should create a regression test for the bug.
- [ ] [@thesahindia] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to the [appropriate location](https://github.com/Expensify/App/blob/main/contributingGuides/TestRail_Navigation_C%2B.md) to ensure the same bug will not reach production again.
- [ ] [@mallenexpensify] Review the proposed regression test steps and location.
- [ ] [@mallenexpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by MelvinBot** [comment]
Looks like something related to `react-navigation` may have been mentioned in this issue discussion.

As a reminder, please make sure that all proposals are not workarounds and that any and all attempt to fix the issue holistically have been made before proceeding with a solution. Proposals to change our `DeprecatedCustomActions.js` files should not be accepted.

Feel free to drop a note in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q) with any questions.

**Comment by mallenexpensify** [comment]
@thesahindia and @priya-zha, can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~01559df482f2b25fe9

**Comment by priya-zha** [comment]
@mallenexpensify I've accepted the offer. Thank you

**Comment by MelvinBot** [comment]
📣 @priya-zha! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by priya-zha** [comment]
Contributor details
Your Expensify account email: pj35134@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~011e32317dbbd489e0

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by mallenexpensify** [comment]
@fedirjh is there an easy way to show the similar/same code that was used?  Trying to think of a way to process-ize situations like this (even though we hope they're rare edge cases)

**Comment by mallenexpensify** [comment]
Thanks @fedirjh , it does appear they are very similar.  I think the reason I stated that there weren't any previous proposals is because I searched for `Proposal` at the beginning of posts and didn't see [yours](https://github.com/Expensify/App/issues/14644#issuecomment-1425089324). Please try to always start any proposals with `## Proposal` 
 
@thesahindia , do you think that @BeeMargarida 's PR include a significant amount of code from @fedirjh 's proposal? 

Also.. side note, we've [updated the CONTRIBUTING.md](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#for-external-agencies-that-expensify-partners-with) to (hopefully) prevent similar circumstances as this in the future. 

> For external agencies that Expensify partners with
> Follow all the above above steps and processes. When you find a job you'd like to work on:
> 
> - Post “I’m from [agency], I’d like to work on this job”
>   - If no proposals have been submitted by other contributors, BugZero (BZ) team member or an internal engineer will assign the issue to you.
>   - If there are existing proposals, BZ will put the issue on hold. Contributor+ (C+) will review the existing proposals. If a contributor’s proposal is accepted then contributor will be assigned to the issue. If not the issue will be assigned to the agency-employee.
> - Once assigned follow the steps [here](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#propose-a-solution-for-the-job) to submit your proposal

**Comment by mallenexpensify** [comment]
Thanks @thesahindia .   @fedirjh , I hired you in Upwork, pay is the job amount, $1000. can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~01559df482f2b25fe9

**Comment by mallenexpensify** [comment]
@fedirjh paid, thanks!

@BeeMargarida @thesahindia @francoisl , can one of you please complete the regression test steps [HERE](https://github.com/Expensify/App/issues/14644#issuecomment-1467263293)?   Please comment if you think the below aren't accurate test steps for the test case?  I created [this GH](https://github.com/Expensify/Expensify/issues/272970) for it already.  

1. Open Expensify in the browser and in iOS (or Android) using the same account
2. Go to the same chat in both the browser and the device and open the emoji popover
3. In the browser, select a skin tone and check the selected skin tone in the device
4. Verify that only the skin tone selected in the browser appears highlighted in the device
5. Verify that no errors appear in the JS console

**Comment by francoisl** [comment]
Thanks, the proposed steps look good to me as well. I updated the regression test steps comment.

**Comment by mallenexpensify** [comment]
whoohoooo.. Everyone's paid, the checklist is complete and the TestRail GH is created, closing!
